enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,294 @@
+      ...
+```
+
+This information can be useful when debugging a stalled operator uninstall process to see the CRs whose removal is blocking operand cleanup.
+
+**NOTE:** The number of CRs that OLM has to list in the status is unbounded and could potentially cause the CSV size to exceed the [max request size][max-request-size] configured in etcd. For this reason there would have to be an upper limit on the number of CRs displayed in `status.cleanup.pendingDeletion`.
+
+The CSV can also be updated with a status condition to indicate that cleanup is in progress when the CSV is blocked on deletion.
+
+```YAML
+status:
+  conditions:
+  - lastTransitionTime: "..."
+    lastUpdateTime: "..."
+    message: |
+      waiting for operator to finish cleanup for 4 CRs
+    phase: Cleanup
+    reason: CleanupWaiting
+```
+
+This CSV will transition to a new phase `Cleanup` with the reason `CleanupWaiting` when it is pending deletion with the finalizer `operatorframework.io/cleanup-apis` present.
+
+#### Handling CSV deletion due to replacement
+
+A CSV can also be [deleted by OLM][csv-deletion-on-upgrade] during it's lifecycle when OLM detects a newer CSV that replaces it as part of an upgrade. For an operator that has already opted-in to operand cleanup, this would result in OLM cleaning up all the CRs before removing the old CSV.
+
+To prevent the cleanup of CRs on an upgrade, OLM should ensure that a CSV which has `spec.cleanup.enabled: true` and is marked for replacement (`phase: replacing`), should first be cleared of the `operatorframework.io/cleanup-apis` finalizer and opted-out of cleanup before it is marked for deletion or deleted.
+
+#### CRD removal
+
+While not strictly necessary for the goal of cleaning up operator-created resources, the cleanup process can also include removing the CRDs after the CRs and operator have been removed. The CRD can't always be removed however.
+
+If there are two non-global operators installed of the same type (providing the same APIs), then removing the CRD during the cleanup of one operator would result in the deletion of CRs for the other operator as well.
+
+To avoid this case OLM would have to ensure that it only removes the CRD during cleanup if:
+
+- The operator installed is global
+- The non-global operator is the only installation for that operator type in the cluster, i.e no other CSVs provide the same owned APIs.
+
+## Design Details
+
+### Test Plan
+
+OLM's e2e testing suite should be upgraded to handle the following use cases:
+
+- An operator that is uninstalled with `spec.cleanup` unspecified or `spec.cleanup.enabled: false` preserves the CRs
+- An operator that is uninstalled with `spec.cleanup.enabled: true` removes all CRs
+- An operator that is uninstalled with `spec.cleanup.enabled: true` should display all CRs that are pending deletion in the CSV status.
+- An CSV with `spec.cleanup.enabled: true` when replaced/upgraded by a new CSV should trigger cleanup when the old CSV is deleted.
```

Comment:
need to cover operators installed w/ various operatorgroups:

1) global operators with CRs in multiple namespaces (all CRs should be removed)
2) single namespace operators with CRs in multiple namespaces (only the CRs in the operator's namespace should be removed)
3) multinamespace operator with CRs in multiple namespaces (only the CRs in the operator's target namespaces should be removed)